File: Problem/By section/Search/DFS/Problem 417/problem_417.java
import java.util.*;
class Solution {
    private final int[][] direction = {{-1,0}, {1,0}, {0,1}, {0, -1}};
    
    // Time: O(w * h), Space: O(w * h)
    public List<List<Integer>> pacificAtlantic(int[][] matrix) {
        if(matrix.length == 0) return (new ArrayList<List<Integer>>());
        
        // dfs on each ocean and if a place can be reached by both ocean
        // add to the ans
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        boolean[][] Atlantic = new boolean[matrix.length][matrix[0].length];
        boolean[][] Pacific = new boolean[matrix.length][matrix[0].length];
        
        for(int i = 0; i < matrix.length; i ++){
            DFS(matrix, i, 0, Pacific);
            DFS(matrix, i, matrix[0].length - 1, Atlantic);
        }
        
        for(int i = 0; i < matrix[0].length; i++){
            DFS(matrix, 0, i, Pacific);
            DFS(matrix, matrix.length - 1, i, Atlantic);
        }
        
        for(int i = 0; i < matrix.length; i ++){
            for(int j = 0; j < matrix[0].length; j++){
                if(Atlantic[i][j] && Pacific[i][j]){
                    ans.add(Arrays.asList(i,j));
                }
            }
        }
        
        return ans;
    }
    
    public void DFS(int[][] matrix, int i, int j, boolean[][] is_Visited){
        is_Visited[i][j] = true;
        for(int[] action : direction){
            int new_x = i + action[0];
            int new_y = j + action[1];
            
            if(new_x < 0 || new_x > matrix.length - 1 || 
               new_y < 0 || new_y > matrix[0].length - 1 || 
               is_Visited[new_x][new_y] || 
               matrix[new_x][new_y] < matrix[i][j]){
                   
                continue;
            }
            
            DFS(matrix, new_x, new_y, is_Visited);

        }
    }
}


public class problem_417{
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[][] test = {{1,2,2,3,5},{3,2,3,4,4},{2,4,5,3,1},{6,7,1,4,5},{5,1,1,2,4}};
        sol.pacificAtlantic(test);
    }
}

File: Problem/By section/Search/DFS/Problem 417/problem_417.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    const vector<vector<int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    // Time: O(n), Space: O(n)
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int m = heights.size();
        if(!m) return vector<vector<int>>{};

        int n = heights[0].size();

        unordered_set<int> Pacific;
        unordered_set<int> Atlantic;

        for(int i = 0; i < m; i++) {
            dfs(heights, Pacific, i, 0, m, n);
            dfs(heights, Atlantic, i, n - 1, m, n);
        }

        for(int i = 0; i < n; i++) {
            dfs(heights, Pacific, 0, i, m, n);
            dfs(heights, Atlantic, m - 1, i, m, n);
        }

        vector<vector<int>> result;
        for(auto& i : Pacific) {
            if(Atlantic.count(i)) {
                int y = i % n;
                int x = (int)((i - y) / n);
                result.push_back({x, y});
            }
        }

        return result;
    }

    bool check_valid(unordered_set<int>& record, int x, int y, int m, int n) {
        int idx = x * n + y;
        return (x < 0 || y < 0 || x >= m || y >= n || record.count(idx));
    }

    void dfs(vector<vector<int>>& heights, unordered_set<int>& record, int x, int y, int m, int n){
        int idx = x * n + y;

        record.insert(idx);

        for(auto& direction : directions) {
            int new_x = x + direction[0];
            int new_y = y + direction[1];

            if(check_valid(record, new_x, new_y, m, n) || heights[x][y] > heights[new_x][new_y]) continue;

            dfs(heights, record, new_x, new_y, m, n);
        }
    }
};

int main(int argc, char** argv) {
    Solution s;
    
    return 0;
}